### PR TITLE
Issue #917: Adds aggregators to regex substring search methods

### PR DESCRIPTION
**Pull request by pierce314159**
This PR (Issue #917):
- Adds aggregators to the regex substring search functions with the goal of improving performance

### CODE REVIEW

File: src/SegmentedArray.chpl [hunk 1]
```diff
@@ -507,31 +507,31 @@ module SegmentedArray {
 
       select mode {
         when SearchMode.contains {
-          forall (o, l, h) in zip(oa, lengths, hits) with (var myRegex = _unsafeCompileRegex(pattern)) {
+          forall (o, l, h) in zip(oa, lengths, hits) with (var myRegex = _unsafeCompileRegex(pattern), var agg = newDstAggregator(bool)) {
             // regexp.search searches the receiving string for matches at any offset
-            h = myRegex.search(interpretAsString(va[o..#l])).matched;
+            agg.copy(h, myRegex.search(interpretAsString(va[o..#l])).matched);
```
Comment (reuster986) [inline src/SegmentedArray.chpl:512]: The way this `forall` and `zip` is structured I believe already guarantees that `h` is local, so I don't think aggregation will help here, although @ronawho might know something I don't.

I'm guessing the biggest slowdown is happening inside of `interpretAsString`, which copies the slice `va[o..#l]` into a local buffer. Most of the time, this copy is unnecessary, so there's a lot of wasted work (this function was originally intended for getting a handful of strings to display back to the user, not for bulk operations, which is why it hasn't been optimized). Unfortunately, there's no clear way to use aggregation in that function either. We might have to do something like what @ronawho did here https://github.com/Bears-R-Us/arkouda/blob/36497dff3b55e21a14edfa16ecff1ea738423756/src/SipHash.chpl#L81-L105

When `startLocale == endLocale == hereLocale`, I think you can do
```chpl
myRegex.search(createStringWithBorrowedBuffer(c_ptrTo(va[o]), l-1, l))
```

with no extra copying. For the second case (string is all on one locale, but not `here`), you could localize with `GET` as above, then use the same `createStringWithBorrowedBuffer` call.

For the final case of a string spanning locales, you would still have to use `interpretAsString`, but there are at most `numLocales-1` of these, so it shouldn't hurt performance much.

**Review by reuster986 — COMMENTED** [review]
See inline comment for strategic suggestion.

File: src/SegmentedArray.chpl (hunk 1 above)
Comment (ronawho) [reply src/SegmentedArray.chpl:512]: I agree that most of the overhead is from the slicing and copy inside `interpretAsString`. https://github.com/Bears-R-Us/arkouda/issues/917 has some more discussion of this. I wasn't sure if `h` was local so I wasn't sure if this PR would have an impact and to start I was just looking at the local overheads.

I think we'll end up with something similar to that localizing code in SipHash, though I'm hoping to move that to a common helper that we can use in both locations.

Comment (pierce314159) [reply src/SegmentedArray.chpl:512]: Awesome! Thanks for the suggestion! I'll try that out

Comment (reuster986) [reply src/SegmentedArray.chpl:512]: @ronawho those are some encouraging results on #917, sorry I missed that discussion. 

That leaves two questions:
- What should we do about this PR and the use of aggregators? I think `h` is always local, but I could be wrong... should we merge this PR to see what effect it has on the benchmark?
- Who wants to write the common helper function?

Comment (ronawho) [reply src/SegmentedArray.chpl:512]: I'd be inclined to make the local optimizations in #917, and see what type of impact that has on multi-locale results. If multi-locale perf still looks poor then I'd explore aggregation or other comm optimizations, but I'd probably close this until then (looks like @pierce314159 has already done so)

The code is a little tricky and low-level so I'm happy to own writing the common helper function. Things are a bit busy here with the upcoming 1.25 release, so it'll probably be next week till I have time to get to that.

Comment (reuster986) [reply src/SegmentedArray.chpl:512]: @ronawho sounds good; totally understand the need to focus on 1.25 right now.

@pierce314159 does that sound ok to you?

Comment (pierce314159) [reply src/SegmentedArray.chpl:512]: sounds good to me!

**Pull request closed without merging**